build-system/tasks/visual-diff.js

```diff
@@ -35,10 +100,463 @@ function setPercyBranch() {
+    log('verbose', 'Setting AMP experiments',
+        colors.cyan(experiments.join(', ')));
+    await page.setCookie({
+      name: 'AMP_EXP',
+      value: experiments.join('%2C'),
+      domain: HOST,
+    });
+  }
+}
+
+/**
+ * Clears all AMP experiment cookies.
+ *
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ */
+async function clearExperiments(page) {
+  await page.deleteCookie({name: 'AMP_EXP', domain: HOST});
+}
+
+/**
+ * Enables debugging if requested via command line.
+ */
+function setDebuggingLevel() {
+  process.env.WEBSERVER_QUIET = '--quiet';
+
+  if (argv.debug) {
+    // eslint-disable-next-line google-camelcase/google-camelcase
+    argv.chrome_debug = true;
+    process.env.WEBSERVER_QUIET = '';
+  }
+  if (argv.webserver_debug) {
+    process.env.WEBSERVER_QUIET = '';
+  }
+}
+
+/**
+ * Creates a Percy build with only a blank page for comparison.
+ *
+ * Enables us to require percy checks on GitHub, and yet, not have to do a full
+ * build for every PR.
+ *
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ */
+async function createEmptyBuild(page) {
+  log('info', 'Skipping visual diff tests and generating a blank Percy build');
+  const blankAssetsDir = '../../examples/visual-tests/blank-page';
+  const percy = new Percy({
+    loaders: [
+      new FileSystemAssetLoader({
+        buildDir: path.resolve(__dirname, blankAssetsDir),
+        mountPath: '',
+      }),
+    ],
+  });
+  await percy.startBuild();
+  await page.goto(`${BASE_URL}/examples/visual-tests/blank-page/blank.html`);
+  await percy.snapshot('Blank page', page, SNAPSHOT_EMPTY_BUILD_OPTIONS);
+  await percy.finalizeBuild();
+}
+
+/**
+ * Simple wrapper around the JS (Percy-Puppeteer) based visual diff tests.
+ */
+async function visualDiffPuppeteer() {
+  if (argv.verify) {
+    const buildId = fs.readFileSync('PERCY_BUILD_ID', 'utf8');
+    const status = await waitForBuildCompletion(buildId);
+    verifyBuildStatus(status, buildId);
+    return;
+  }
+
+  if (!process.env.PERCY_PROJECT || !process.env.PERCY_TOKEN) {
+    log('fatal', 'Could not find', colors.cyan('PERCY_PROJECT'), 'and',
+        colors.cyan('PERCY_TOKEN'), 'environment variables');
+  }
+  setDebuggingLevel();
+
+  // Launch a browser and local web server.
+  const page = await launchBrowser();
+  await launchWebServer().catch(reason => {
+    log('fatal', `Failed to start a web server: ${reason}`);
+  });
+
+  if (argv.skip) {
+    await createEmptyBuild(page);
+    // Explicitly exit, to trigger the webserver's exit event too.
+    process.exit();
+    return;
+  }
+
+  // Load and parse the config. Use JSON5 due to JSON comments in file.
+  const visualTestsConfig = JSON5.parse(
+      fs.readFileSync(
+          path.resolve(__dirname, '../../test/visual-diff/visual-tests'),
+          'utf8'));
+  await runVisualTests(page, visualTestsConfig);
+
+  // Explicitly exit, to trigger the webserver's exit event too.
+  process.exit();
```

Comment:
I recently discovered that you can allow a `node` process to print all its output before exiting. To so, replace...
```
process.exit(code);
```
... with...
```
process.exitCode = code;
```
... and allow the process to exit after its stdout is fully flushed.

Once you do that to all your exit points, the `sleep` should no longer be needed.

Reply:
I played a bit with it but the process hangs unless I explicitly called process.exit() at the end. This works as is and I don't want to mess around with it for too long